AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/ModelSyncedPayload.swift

```diff
@@ -0,0 +1,31 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+struct ModelSyncedPayload {
+    var modelName: String
+    var isFullSync: Bool
+    var isDeltaSync: Bool
+    var createCount: Int
+    var updateCount: Int
+    var deleteCount: Int
+
+    init(modelName: String? = nil,
+         isFullSync: Bool? = nil,
+         isDeltaSync: Bool? = nil,
+         createCount: Int? = nil,
+         updateCount: Int? = nil,
+         deleteCount: Int? = nil) {
+        self.modelName = modelName ?? ""
+        self.isFullSync = isFullSync ?? false
+        self.isDeltaSync = isDeltaSync ?? false
+        self.createCount = createCount ?? 0
+        self.updateCount = updateCount ?? 0
+        self.deleteCount = deleteCount ?? 0
+    }
```

Comment:
1. Simplify this by declaring the defaults in the initializer signature and making the argument types non-optional:

```suggestion
    init(modelName: String = "",
         isFullSync: Bool = false,
         isDeltaSync: Bool = false,
         createCount: Int = 0,
         updateCount: Int = 0,
         deleteCount: Int = 0) {
        self.modelName = modelName
        self.isFullSync = isFullSync
        self.isDeltaSync = isDeltaSync
        self.createCount = createCount
        self.updateCount = updateCount
        self.deleteCount = deleteCount
    }
```

2. But even with this, some of these defaults don't really make sense. When would `modelName` be nil or empty? As noted above, what would it mean to have `isFullSync` and `isDeltaSync` both be true or both be false?

My guess is that some of these decisions are so you can build this incrementally inside the operation. In that case, consider a `Builder` pattern, like I've sketched out here:

```swift
public struct ModelSyncedPayload {
    // For internal use only
    enum SyncType {
        case full
        case incremental
        case unspecified
    }
    let syncType: SyncType

    public let modelName: String

    // To conform with other platform specs, if needed
    public var isFullSync: Bool {
        syncType == .full
    }

    public var isDeltaSync: Bool {
        syncType == .incremental
    }

    public let createCount: Int
    public let updateCount: Int
    public let deleteCount: Int

    public init(
        modelName: String,
        syncType: SyncType,
        createCount: Int,
        updateCount: Int,
        deleteCount: Int
    ) {
        self.modelName = modelName
        self.syncType = syncType
        self.createCount = createCount
        self.updateCount = updateCount
        self.deleteCount = deleteCount
    }
}

extension ModelSyncedPayload {
    struct Builder {
        var modelName: String
        var syncType: SyncType
        var createCount: Int
        var updateCount: Int
        var deleteCount: Int
        init () {
            modelName = ""
            syncType = .unspecified
            createCount = 0
            updateCount = 0
            deleteCount = 0
        }

        func build() -> ModelSyncedPayload {
            ModelSyncedPayload(
                modelName: modelName,
                syncType: syncType,
                createCount: createCount,
                updateCount: updateCount,
                deleteCount: deleteCount
            )
        }
    }
}

func atCallSite() {
    var builder = ModelSyncedPayload.Builder()
    builder.modelName = "Foo"
    builder.createCount += 1
    builder.syncType = .full
    // ...
    // callback(.success(builder.build()))
}
```